### ISSUE TITLE
initial step sizes in SkyModel in 3D analysis

### ISSUE THREAD

**Issue by facero** [gammapy/gammapy#1531]
When performing a 3D fit with a Skymodel, the user has to specify reasonable values for the steps in the iMinuit fitting process. This is currently done via the skymodel.parameters.set_parameter_errors  interface.

 If no values are provided iMinuit assumes that the step is one. This is wrong for all our parameters (and super wrong for the amplitude parameter).
After playing around some values, the following example allows the fit of a Gaussian in DC1 to converge (see detailed example [here ](https://github.com/gammasky/cta-analyses/blob/master/dc-1-checks/rx_j1713/RXJ_gammapy_data_analysis3D.ipynb)on the DC1 analysis repo, cta users only).

```
spatial_model = SkyGaussian(
    lon_0=target_position.ra,
    lat_0=target_position.dec,
    sigma='0.1 deg',
)

spectral_model = PowerLaw(
    index=2.,
    amplitude='1e-12 cm-2 s-1 TeV-1',
    reference='1 TeV',
)

skymodel = SkyModel(
    spatial_model=spatial_model,
    spectral_model=spectral_model,
)

skymodel.parameters.set_parameter_errors({
    'lon_0': '0.001 deg',
    'lat_0': '0.001 deg',
    'sigma': '0.001 deg',
    'index': '0.01',
    'amplitude': '0.01e-12 cm-2 s-1 TeV-1',
})
```

The question here is, do we want to provide some default values to those parameters while leaving the default values still changeable for specific cases.
Most users will never change those values and this will allow the SkyModel specification to be less cumbersome and help fit convergence.

**Comment by cdeil** [comment]
I think leaving this to users is possible; that's what Fermi ST do e.g., they just use the "scale" and "error" and pass that on directly to MINUIT?

But IMO it would be better to handle this by default for the user. Could be done either by a generic scaling of parameters to value 1 in the interface to the optimiser and error estimator, or even by using our domain knowledge that e.g. for RA and DEC if in deg then 0.1 or 0.01 is a good initial step size to use by default.

A little bit related is the model parameter API question: #1398

I've put this on the v0.9 milestone for now, but we can discuss on Friday if this should be super high priority and still go in v0.8, because changes here will affect all users. This is not a small task. We have to discuss who takes care of this within the next ~ week or how to split the work, so that we can improve quickly.

@hdembinski @joleroi @registerrier @woodmd @adonath - Thoughts on how this should be done?

**Comment by adonath** [comment]
I think we should use our domain knowledge to introduce a sensible default step size for the model parameters. This could be attached to the `Parameter` object by a hidden attribute `._minuit_default_stepsize`. When parameter errors are available, those would be used instead. 

The more powerful concept might be to introduce a `scale` attribute for the `Parameter` class as suggested by @cdeil, which might solve the problem of the default step size as well. I think rescaling fit parameters for the optimizer is a good idea in any case. So I actually prefer the latter solution.

**Comment by facero** [comment]
Lessons learnt from Fermi:
- the way this is handled in Fermi and (probably) in ctools is that all parameters are scaled to ~1 in the XML. Minuit doesn't know anything about the physical parameters. 
One important difference is that in Fermi gtlike, the RA, DEC position are not let free. So it's only spectral parameters that are scaled to 1. But how to scale RA, DEC ? 
- When the amplitude of the POW is renormed it is better to fix its min value to zero instead of some small values (e.g. 1e-3)
- it's ideally better to have unbound param space (no min, max) but this could allow negative values. When bounding the parameters, the max should not be too high from the truth value otherwise the  param space really explored will be too small wrt to the defined min-max param space.
- pointlike (the internal science tools) fits Log(norm) instead of a scaled norm and it seems to work well and defacto forbids negative norms.

For all spatial parameters (RA, DEC, sigma, radius, etc) it seems reasonable to me to provide some default values. The scaling is mostly important for spectral param like Norm (but Log(Norm) is worth investigating), Ecut, Ecentroid, etc
Maybe having a hybrid solution for spatial param (some default values) and spectral param (scales?)

**Comment by HDembinski** [comment]
> But IMO it would be better to handle this by default for the user. Could be done either by a generic scaling of parameters to value 1 in the interface to the optimiser and error estimator, or even by using our domain knowledge that e.g. for RA and DEC if in deg then 0.1 or 0.01 is a good initial step size to use by default.

> I think we should use our domain knowledge to introduce a sensible default step size for the model parameters.

- Minuit works more reliably when all fitted parameters of the model are of similar scale. Internally, it does linear algebra that involves summing. In a sum you lose numerical precision when you add numbers which differ at lot in magnitude. If you have parameters of O(1) and O(1000), I wouldn't worry, but I would worry when it is O(1) and O(1e10). It is always possible to scale the parameters in the model so that the magnitudes of the fitted parameters are similar.

- In my experience, a good initial guess for the step size is 0.1 x, when x is the initial parameter value, except if the initial parameter is 0, then you should just use 0.1. We could even change the default in iminuit along these lines, if one could prove with some benchmarks that this is a better choice.

- That being said, I recently did some numerical tests with iminuit on a simple quadratic function, and changing the initial step size did not have a measurable effect on the number of iterations used to find the minimum. You can easily try this out for your problem.

**Comment by cdeil** [comment]
Let's make this issue a focus in our Gammapy call this Friday ([link](https://github.com/gammapy/gammapy-meetings/tree/master/2018-07-20)).

All of the options mentioned are good one, it'll be hard to choose what to bake into Gammapy, it needs some experimentation.

If someone plans to work on this already this week, I think there's two ways to proceed:

1. Add a "scale" in [Parameter](https://github.com/gammapy/gammapy/blob/42416f5c8bd904a27d70ba8607bfe3db582fb6b6/gammapy/utils/modeling.py#L21) and then write a utility function that sets one or different schemes as suggested above, moving things like 1e-11 for flux to the scale and letting the optimiser see something better scaled. It might be non-trivial, especially concerning the scaling back of covar in a good place.
2. Auto-scale in the interface to iminuit, don't add the scale option to Parameter.

I vaguely remember a discussion with @woodmd a year or two ago, where he commented that offering the "scale" for parameters in the Fermi ST for users was a mistake, it would have been better to just re-scale at the optimiser interface. Even if that would be the long-term solution for Gammapy, it might be useful to add a scale to Parameter to allow for easier experimentation for a while.

**Comment by HDembinski** [comment]
In principle, Minuit2 in C++ supports arbitrary transformations between "external" and "internal" parameters. This is currently used to implement bounded parameters, by mapping a parameter that goes from a to b to an internal parameter that goes from -inf to inf. This machinery could be used to also rescale parameters internally. The advantage of doing it on this deep level would be, that you don't have to scale the input and output of Minuit, even in the Minuit prints the parameters would appear in their natural scale.

That being said, hacking C++ Minuit2 is a pain and if you can do it on your end, you get something that works with low latency.

**Comment by cdeil** [comment]
I think we should do such things in Gammapy. Note that we plan to have other fitting backends, notably Sherpa and emcee, and some tricks like e.g. transforming flux to log flux should likely be applied there as well.

@HDembinski - In `iminuit`, I would always start with a docs example, and only if several users request it make code additions. Looks like this is what we currently have:
http://nbviewer.jupyter.org/github/iminuit/iminuit/blob/master/tutorial/tutorial.ipynb#Initial-parameters
I would suggest to make "initial step size" as well as "errordef" separate headings one can link to. Especially "errordef" is really a separate thing from initial parameters.

Another option could be to add one example with a failing fit with parameters of scale 1e-20 and 1, and then showing how to fix this by rescaling, in a new notebook or the advanced tutorial.

**Comment by HDembinski** [comment]
> I think we should do such things in Gammapy. Note that we plan to have other fitting backends, notably Sherpa and emcee, and some tricks like e.g. transforming flux to log flux should likely be applied there as well.

That is an excellent argument, agreed.

Full ack on the suggestions for improving the tutorial and mentioning this issue there.

**Comment by facero** [comment]
So I was struggling to fit the supernova remnant RXJ1713 with a Disk model and the fit always failed to converge. I found the problem: the initial value for the amplitude was too far from the solution. I had amplitude=1e-12 whereas the truth value was 1.5e-11 (the step size was 0.1e-12).

This is probably a good example of why working with very small values is a bad idea. 
In CTA data we will have amplitude ranging from a few 1e-11 down to 1e-13. So having one scale factor might not be enough. Log(norm) might be worth investigating.

**Comment by HDembinski** [comment]
@facero Just to clarify, if all your parameters are O(1e-12), that's ok. It is bad if you have an amplitude with a scale 1e-12 and a gamma exponent which is of O(1). The problem only occurs when you fit several parameters simultaneously. When you only fit one parameter, don't worry about its scale.

When you change your fit to use log(norm) instead of norm, it helps with the convergence issues. I used this technique myself and know from experience and it helps with the convergence. However, if you change a parameter from norm to log(norm), you may also introduce a bias on the fitted parameter.

This is a bit cumbersome to explain. Let's assume that a fit of a parameter x is unbiased. Unbiased  means that if you repeat the fit with independent data many times, the average of all the fitted x values converges to the true value. Now we know that log(<x>) != <log(x)>, because taking the avergae <...> is a linear operation, but the log(...) function is non-linear. The two do not commute. Therefore, if you fit log(x) instead of x, log(x) must be biased when x is not biased.

The best solution is to use good starting values, or to use log(norm) as a parameter first for converge and then repeat the fit with a norm parameter, using the log(norm) result as a starting point. I wrote a fit called GSF of all nuclear cosmic ray data where I employed several tricks to make the fit converge reliably without introducing bias. https://pos.sissa.it/301/533/

**Comment by facero** [comment]
@HDembinski  thanks for the summary it was very clear. 
As you say the problem in my case is that I'm trying to fit multiple parameters which have very different scales. In my example I have : 
- RA ~ 258
- DEC~ -39
- radius~0.1-1
- POW index ~2-3
- amplitude ~ 1e-13 - 1e-11

so quite messy indeed.

**Comment by cdeil** [comment]
@joleroi agreed to hack on this today.

**Comment by HDembinski** [comment]
@cdeil I think the float division by zero is a bug in the new iminuit. I think I got the same error in a completely different fit. I am looking into this.

**Comment by HDembinski** [comment]
I made a PR with a fix for iminuit. As soon as the PR is merged, you can try with the iminuit master, the problem should be gone then.

**Comment by cdeil** [comment]
Is it https://github.com/iminuit/iminuit/pull/284 ? Does this run during the `hesse` call?

**Comment by HDembinski** [comment]
Like we discussed offline, this error happened later. You probably ran with print_level = 1 or higher, so that iminuit tried to be helpful and print the correlation matrix at the end of the hesse call. That triggered the bug.

**Comment by cdeil** [comment]
Should be fixed via #1696 .

Of course, there might be bugs, and better ways to improve fit robustness and speed (fewer function evaluations) will remain a topic for a long time.

But I'm closing this issue now. Please open new ones when you find issues (and reference this one if you like)!
